src/Libraries/Microsoft.Extensions.AI.Abstractions/Embeddings/BinaryEmbedding.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections;
+using System.ComponentModel;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>Represents an embedding composed of a bit vector.</summary>
+public sealed class BinaryEmbedding : Embedding
+{
+    /// <summary>The embedding vector this embedding represents.</summary>
+    private BitArray _vector;
+
+    /// <summary>Initializes a new instance of the <see cref="BinaryEmbedding"/> class with the embedding vector.</summary>
+    /// <param name="vector">The embedding vector this embedding represents.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="vector"/> is <see langword="null"/>.</exception>
+    public BinaryEmbedding(BitArray vector)
+    {
+        _vector = Throw.IfNull(vector);
+    }
+
+    /// <summary>Gets or sets the embedding vector this embedding represents.</summary>
+    [JsonConverter(typeof(VectorConverter))]
+    public BitArray Vector
+    {
+        get => _vector;
+        set => _vector = Throw.IfNull(value);
+    }
+
+    /// <inheritdoc />
+    [JsonIgnore]
+    public override int Dimensions => _vector.Length;
+
+    /// <summary>Provides a <see cref="JsonConverter{BitArray}"/> for serializing <see cref="BitArray"/> instances.</summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class VectorConverter : JsonConverter<BitArray>
```

Comment:
A fully serialized embedding would look as follows:
```json
{
   "vector": { "length": xxx, "bits": xxx }
}
```
Not sure if I'm a fan of the nestedness here. If it's possible to make the embedding value immutable, perhaps one alternative would be to

1. Hide the BitArray behind a read-only indexer and
2. Expose a pair of `Length` and `Bits` properties on the embedding itself (perhaps with EB.Never annotation) that map from the input `BitArray`.

Reply:
Should we just have a bitstring for the JSON representation, i.e. `"vector": "01101010111"`, rather than length + base64?

Reply:
> If it's possible to make the embedding value immutable

The embedding types are mutable. I don't see us changing that now.

> Should we just have a bitstring for the JSON representation, i.e. "vector": "01101010111", rather than length + base64?

I went with base64 to keep it smaller, e.g. a 1024 bit vector would be ~170 characters with base64 vs 1024 characters with such a character-per-bit encoding. But if we don't care about that, sure, we could do "0110..." or "[0, 1, 1, 0, ...]" to match other types.